src/planning.rs

```diff
@@ -100,12 +100,10 @@ impl<'a> BuildPlanner<'a> {
 
     let mut crate_contexts = Vec::new();
 
-    let source_id = match self.registry.clone() {
-      Some(v) => v,
-      None => try!(SourceId::crates_io(&self.cargo_config)),
-    };
-
-    let package_ids = try!(find_all_package_ids(source_id, &resolve));
+    let mut package_ids: Vec<PackageId> = Vec::new();
+    for dep in root_direct_deps.iter() {
+      package_ids.extend(find_all_package_ids(dep.source_id().clone(), &resolve)?);
```

Comment:
This might be better implemented by changing `find_all_package_ids` to take the root_package_id and filter out only that package_id. The current implementation takes the source_id only to avoid including the root (kinda cludgey, given that we have the root's package_id available, I know).

In any case, this implementation appears to loop over the whole dep set way more times than it probably needs to, given its objective of supporting git deps.

_Addendum_: This extra looping also causes a minor breakage: cargo.bzl and the alias BUILD file end up with one duplicate per dependency (so 5 deps caused there to be five copies of each dependency). Test suite needs help :cry: (that https://github.com/google/cargo-raze/issues/7 will begin to deliver)

Reply:
I think what you're ultimately after is some way of deduplicating `package_ids` after the iteration over root deps?
The loop was put in place because it originally assumed a `SourceId` of crates.io, and this now fixes is so that each root dependency uses its own Source (adding support for non-crates.io crates in addition to Git).
Happy to write up some test cases, wanted to get a commit started to solicit feedback like this!